lib/python/picongpu/extra/input/preparingInsightData.py

```diff
@@ -0,0 +1,472 @@
+        )  # correcting near field
+        self.Ew = np.abs(self.Ew) * np.exp(1j * (np.angle(self.Ew) - phase_centr + dispersion))  # correcting far field
+
+        self.isPhaseCorrected = True
+        print(
+            "Phase has been corrected. \nApplied dispersion parameters: GVD = %.f fs**2/rad, TOD = %.f fs**3/rad**2"
+            % (GVD, TOD)
+        )
+
+    def correctUglySpotInNF(self, x_ugly, y_ugly, uglybins=3, method="linear"):
+        """
+        In case there is an ugly spot in the near field (such as an unnatural peak), it can be smoothened out with this function.
+        It automatically also corrects the far field by propagating the corrected near field back into the focus.
+
+        Arguments:
+        x_ugly: (near field) x coordinate of the ugly spot
+        y_ugly: (near field) y coordinate of the ugly spot
+        uglybins: number of bins surrounding the ugly spot which will be cut away / smoothened out (default is 3)
+        method: RegularGridInterpolator interpolation method,
+                default is 'linear' (since 'cubic' can cause some bumps in amplitude or phase)
+        """
+        # index borders of hole
+        x_ugly_idx = np.abs(self.x_NF - x_ugly).argmin()
+        y_ugly_idx = np.abs(self.y_NF - y_ugly).argmin()
+        xidx_low = x_ugly_idx - uglybins
+        xidx_upp = x_ugly_idx + uglybins
+        yidx_low = y_ugly_idx - uglybins
+        yidx_upp = y_ugly_idx + uglybins
+
+        x_surr = np.concatenate((self.x_NF[xidx_low - 3 : xidx_low], self.x_NF[x_ugly_idx + uglybins : xidx_upp + 3]))
+        Ew_surr = np.concatenate(
+            (
+                self.Ew_NF[yidx_low:yidx_upp, xidx_low - 3 : xidx_low, :],
+                self.Ew_NF[yidx_low:yidx_upp, xidx_upp : xidx_upp + 3, :],
+            ),
+            axis=1,
+        )
+        interp = RegularGridInterpolator((self.y_NF[yidx_low:yidx_upp], x_surr, self.w), Ew_surr, method=method)
+        Y_hole, X_hole, W_hole = np.meshgrid(
+            self.y_NF[yidx_low:yidx_upp], self.x_NF[xidx_low:xidx_upp], self.w, indexing="ij"
+        )
+        self.Ew_NF[yidx_low:yidx_upp, xidx_low:xidx_upp, :] = interp((Y_hole, X_hole, W_hole))
+
+        # correct the far field
+        preproutines.NFtoFF(self)
+
+        print("corrected ugly spot in near field at x = %.2f mm, y = %.2f mm" % (x_ugly, y_ugly))
+
+    def shiftNFtoCenter(self):
+        """
+        If the near field is not centered around (0, 0), it can be done with this function.
+        It automatically also corrects the far field by propagating the centered near field back into the focus.
+        """
+        nx = int(self.xc_NF / self.dx_NF + 0.5)  # number of pixels to shift in x-direction
+        ny = int(self.yc_NF / self.dy_NF + 0.5)  # number of pixels to shift in y-direction
+
+        if nx == 0 and ny == 0:
+            print("already centered, no corrections necessary")
+            return 0
```

Comment:
You should probably check whether `nx` and `ny` are within half the size of the data extent. Otherwise you just shift the pulse from one half of the screen to the other, don't you?